File: services/graph4j_service/src/process/src/main/java/process/pubsub/UserInitPublisher.java
package process.pubsub;

import process.network.Envelope;

public class UserInitPublisher extends Publisher {
    private final String socketId;
    private final String algorithmName;

    public UserInitPublisher(String socketId, String algorithmName) {
        super(System.getenv("USER_METRICS_TOPIC"), socketId + "_user_metric", "direct");
        this.socketId = socketId;
        this.algorithmName = algorithmName;
    }

    public void send(String time) {
        String envelope = Envelope.sendUserInitTime(socketId, algorithmName, time);
        super.send(envelope);
    }
}

File: services/networkx_service/src/process/pubsub/user_init_publisher.py
import os
import json
from .publisher import Publisher
from network.envelope import Envelope


class UserInitPublisher(Publisher):
    def __init__(self, socket_id, algorithm_name):
        user_metrics_topic = os.environ["USER_METRICS_TOPIC"]
        routing_key = f"{socket_id}_user_metric"
        super().__init__(user_metrics_topic, routing_key, 'direct')

        self.socket_id = socket_id
        self.algorithm_name = algorithm_name

    def send(self, time):
        envelope = Envelope.send_user_init_time(self.socket_id, self.algorithm_name, time)
        super().send(json.dumps(envelope), 0)
